components/layout/query.rs

```diff
@@ -45,9 +46,92 @@ use crate::flow::inline::construct::{TextTransformation, WhitespaceCollapse};
+    /// Get the scroll offset of a viewport (i.e. window).
+    pub root_scroll_offset_getter: &'dom dyn Fn() -> Vector2D<Au>,
+    /// Get the scroll offset of a BoxFragment.
+    pub scroll_offset_getter: &'dom dyn Fn(&BoxFragment) -> Vector2D<Au>,
+}
+
+impl PostCompositeQueryHelper<'_> {
+    fn retrieve_box_fragment(fragment: &Fragment) -> Option<&ArcRefCell<BoxFragment>> {
+        match fragment {
+            Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => Some(box_fragment),
+            _ => None,
+        }
+    }
+
+    /// Get the position property for current node. It returns None for a [`LayoutNode`]
+    /// that doesn't have a [`Fragment`] that correspond to it (e.g. 'display: contents').
+    fn retrieve_position_property(node: ServoLayoutNode<'_>) -> Option<Position> {
+        node.fragments_for_pseudo(None)
+            .first()
+            .and_then(PostCompositeQueryHelper::retrieve_box_fragment)
+            .map(|fragment| fragment.borrow().style.get_box().position)
+    }
+
+    /// Find the parent node in containing block chain, ignoring the nodes that
+    /// does not generate boxes.
+    fn parent_containing_block(
+        node: ServoLayoutNode<'_>,
+        position: Position,
+    ) -> Option<ServoLayoutNode> {
+        fn parent_establishes_containing_block_for_position(
+            parent_style: &ServoArc<ComputedValues>,
+            parent_fragment_flags: FragmentFlags,
+            descendant_position: Position,
+        ) -> bool {
+            match descendant_position {
+                Position::Absolute => parent_style
+                    .establishes_containing_block_for_absolute_descendants(parent_fragment_flags),
+                Position::Fixed => parent_style
+                    .establishes_containing_block_for_all_descendants(parent_fragment_flags),
+                _ => true,
+            }
+        }
+
+        let mut maybe_parent_node = node.parent_node();
+        while let Some(parent_node) = maybe_parent_node {
+            if let Some(parent_fragment) = parent_node.fragments_for_pseudo(None).first() {
+                let established_containing_block = match parent_fragment {
+                    Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => {
+                        parent_establishes_containing_block_for_position(
+                            &box_fragment.borrow().style,
+                            box_fragment.borrow().base.flags,
+                            position,
+                        )
+                    },
+                    Fragment::Positioning(fragment) => {
+                        fragment.borrow().style.as_ref().is_some_and(|style| {
+                            parent_establishes_containing_block_for_position(
+                                style,
+                                fragment.borrow().base.flags,
+                                position,
+                            )
+                        })
```

Comment:
```suggestion
                            parent_establishes_containing_block_for_position(
                                &fragment.borrow().style,
                                fragment.borrow().base.flags,
                                position,
                            )
```
Need to resolve this conflict due to latest incremental layout change #36978